File: CPP/Algorithm/LIS_(Longest_Increasing_Subsequence).cpp
// Note: Gives result of STRICTLY incresing subsequence
// For longest non-decreasing subsequence, replace 'lower_bound' with 'upper_bound' and for longest decreasing subsequence, replace aa[i] *= -1;
// Iterative Approach to determine the length: O(n * log(n))

int64_t aa[N], n;
int lis() {
    vector<int64_t> ans; ans.push_back(aa[0]);
    for (int i = 1; i < n; ++i) {
        if (aa[i] > ans.back()) ans.push_back(aa[i]);
        else {
            int low = lower_bound(ans.begin(), ans.end(), aa[i]) - ans.begin();
            ans[low] = aa[i];
        }
    }
    return (int)ans.size();
}

// Complexity: O(n^2)  // For all of the following codes

int lis(auto &v) {
    int n = v.size(), dp[n], ans = 1;
    for (int i = 0; i < n; ++i) {
        dp[i] = 1;
        for (int j = 0; j < i; ++j) {
            if (v[j] < v[i]) ans = max(ans, dp[i] = max(dp[i], dp[j] + 1));
        }
    }
    return ans;
}

// Operation:
    cin >> n;
    vector<int64_t> v(n); for (i = 0; i < n; ++i) { cin >> v[i]; }
    cout << lis(v);

// Recursive Approach:

int aa[N], dp[N];
int lis(int i) {
    if (dp[i] != -1) return dp[i];
    int ans = 1;
    for (int j = 0; j < i; ++j) {
        if (aa[i] > aa[j]) ans = max(ans, lis(j) + 1);
    }
    return dp[i] = ans;
}

// Operation:
    cin >> n; for (i = 0; i < n; ++i) cin >> aa[i];
    memset(dp, -1, sizeof dp);
    int ans = 0;
    for (i = 0; i < n; ++i) ans = max(ans, lis(i));
    cout << ans;

// To determine the subsequence:

int aa[N], dp[N], in = 0; vector<int> sub;
int lis(auto i) {
    if (dp[i] != -1) return dp[i];
    int ans = 1; bool ok = true;
    for (int j = 0; j < i; ++j) {
        if (aa[i] >= aa[j]) {
            ans = max(ans, lis(j) + 1);
            if (ok) {
                sub.emplace_back(aa[i]);
                in = i; ok = false;
            }
        }
    }
    return dp[i] = ans;
}

// Operation:
    cin >> n; for (i = 0; i < n; ++i) cin >> aa[i];
    memset(dp, -1, sizeof dp);
    for (i = 0; i < n; ++i) lis(i);
    for (i = 0; i < in; ++i) {
        if (aa[i] <= aa[in]) {
            sub.emplace_back(aa[i]); break;
        }
    }
    cout << sub.size() << '\n';
    for (i = sub.size() - 1; i >= 0; --i) cout << sub[i] << ' ';
    // reverse(sub.begin(), sub.end());
    // for (auto &e : sub) cout << e << ' ';

File: Python/Algorithm/LIS_(Longest_Increasing_Subsequence).py
# Note: Gives result of STRICTLY incresing seq
# For longest non-decreasing seq, replace 'lower_bound' with 'upper_bound' and for longest decreasing seq, replace ar[i]*=-1;
# To determine the length:
# Iterative Approach:
# Complexity: O(n * log(n))

def lis(arr):
    from bisect import bisect_left
    ans = [arr[0]]
    for i in range(1, len(arr)):
        if arr[i] > ans[-1]: ans.append(arr[i])
        else:
            low = bisect_left(ans, arr[i])
            ans[low] = arr[i]
    return len(ans)

# Complexity: O(n ^ 2)
# DP Approach:

def lis(v):
    dp = [1] * len(v)
    for i in range(len(v)):
        for j in range(i):
            if v[j] < v[i]: dp[i] = max(dp[i], dp[j] + 1)
    return max(dp)

# Recursive Approach:

def rec(i, v, dp):
    if dp[i] != -1: return dp[i]
    ans = 1
    for j in range(i):
        if v[i] > v[j]: ans = max(ans, rec(j, v, dp) + 1)
    dp[i] = ans; return ans

def lis(v):
    n, ans = len(v), 0; dp = [-1] * n
    for i in range(n):
        ans = max(ans, rec(i, v, dp))
    return ans

# To determine the sequence:

def lis(arr):
    dp = [1] * len(arr); prev = [-1] * len(arr)
    for i in range(len(arr)):
        for j in range(i):
            if arr[j] < arr[i] and dp[i] < dp[j] + 1:
                dp[i] = dp[j] + 1; prev[i] = j
    idx = dp.index(max(dp)); seq = []
    while idx != -1:
        seq.append(arr[idx]); idx = prev[idx]
    seq.reverse(); return seq